test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/SparseFieldSets/SparseFieldSetTests.cs

```diff
@@ -0,0 +1,273 @@
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Bogus;
+using FluentAssertions;
+using JsonApiDotNetCore.Configuration;
+using JsonApiDotNetCore.Serialization.Objects;
+using JsonApiDotNetCore.Services;
+using JsonApiDotNetCoreMongoDbExample;
+using JsonApiDotNetCoreMongoDbExample.Models;
+using Microsoft.Extensions.DependencyInjection;
+using MongoDB.Driver;
+using Xunit;
+
+namespace JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.SparseFieldSets
+{
+    public sealed class SparseFieldSetTests : IClassFixture<IntegrationTestContext<Startup>>
+    {
+        private readonly IntegrationTestContext<Startup> _testContext;
+        private readonly Faker<User> _userFaker;
+
+        public SparseFieldSetTests(IntegrationTestContext<Startup> testContext)
+        {
+            _testContext = testContext;
+
+            testContext.ConfigureServicesAfterStartup(services =>
+            {
+                services.AddSingleton<ResourceCaptureStore>();
+
+                services.AddResourceRepository<ResultCapturingRepository<Blog>>();
+                services.AddResourceRepository<ResultCapturingRepository<Article>>();
+                services.AddResourceRepository<ResultCapturingRepository<Author>>();
+                services.AddResourceRepository<ResultCapturingRepository<TodoItem>>();
+
+                services.AddScoped<IResourceService<Article, string>, JsonApiResourceService<Article, string>>();
+            });
+            
+            _userFaker = new Faker<User>()
+                .RuleFor(u => u.UserName, f => f.Internet.UserName())
+                .RuleFor(u => u.Password, f => f.Internet.Password());
+        }
+
+        [Fact]
+        public async Task Can_select_fields_in_primary_resources()
+        {
+            // Arrange
+            var store = _testContext.Factory.Services.GetRequiredService<ResourceCaptureStore>();
+            store.Clear();
+
+            var article = new Article
+            {
+                Caption = "One",
+                Url = "https://one.domain.com"
+            };
+
+            await _testContext.RunOnDatabaseAsync(async db =>
+            {
+                var collection = db.GetCollection<Article>(nameof(Article));
+                await collection.DeleteManyAsync(Builders<Article>.Filter.Empty);
+                await collection.InsertOneAsync(article);
+            });
+
+            var route = "/api/v1/articles?fields[articles]=caption"; // TODO: once relationships are implemented select author field too
```

Comment:
The original tests are:
- Can_select_fields_in_primary_resources: one attribute + one relationship
- Can_select_attribute_in_primary_resources: only one attribute
- Can_select_relationship_in_primary_resources: only one relationship
- ...
- Can_select_fields_in_primary_resource_by_ID

So I think these should become:
- Cannot_select_fields_with_relationship_in_primary_resources
- Can_select_attribute_in_primary_resources
- Cannot_select_relationship_in_primary_resources
- ...
- Can_select_attribute_in_primary_resource_by_ID

Also add negatives for Can_select_fields_of_HasOne_relationship, Can_select_fields_of_HasMany_relationship, Can_select_fields_of_HasManyThrough_relationship.

Remove these (query string validations):
- Cannot_select_on_unknown_resource_type
- Cannot_select_attribute_with_blocked_capability

Reply:
For now I've only removed the tests you said to remove. 

I don't quite understand what I should do about the rest... I mean, I do understand having to add negatives for relationships but not the rest. Can you explain this?

Reply:
The way I look at the tests is by comparing them against the main repo. If you want to see what I see, use [Araxis Merge](https://www.araxis.com/merge/index.en), which is the best diff/merge tool I've ever encountered. There's a 30-day trial available.

So what I'm proposing is to take everything from main, then apply the mentioned changes. So (...) means: unchanged. Hope this clarifies my intentions.

Reply:
Right now I've just added the tests for relationships.

Reply:
@bart-degreed I was just trying to bring in the `Cannot_select_fields_with_relationship_in_primary_resources` test and it turns out that the url `/api/v1/articles?fields[articles]=caption,author` does not fail... The `QueryLayer` has a projection with only two fields, them being `id` and `caption` but no `author` in sight so the validator cannot do its job here. I checked that `Author` was declared as a `HasOne` relationship and it is. Any ideas?

Reply:
Hmm... yes, this is correct. Because there is no `include`, the relationship is not fetched. However, putting `author` in the fieldset results in rendering the relationship link (without any data in it).

See the note "Pay attention to the fact that you have to add..." at https://jsonapi.org/examples

I haven't tried, but for your test to fail, I think you need to add the following code to `MongoDbQueryExpressionValidator`:
```c#
public override QueryExpression VisitSparseFieldSet(SparseFieldSetExpression expression, object argument)
{
    if (expression != null && expression.Fields.Any(field => field is RelationshipAttribute))
    {
        throw new UnsupportedRelationshipException();
    }

    return base.VisitSparseFieldSet(expression, argument);
}
```

Reply:
Oh well, that does not work, because relationships in sparse fieldsets do not propagate up till there. Instead, add the following to `MongoDbRepository.ApplyQueryLayer`:
```c#
bool hasRelationshipSelectors = _constraintProviders
    .SelectMany(p => p.GetConstraints())
    .Select(expressionInScope => expressionInScope.Expression)
    .OfType<SparseFieldTableExpression>()
    .Any(fieldTable => fieldTable.Table.Values.Any(fieldSet => fieldSet.Fields.Any(field => field is RelationshipAttribute)));

if (hasRelationshipSelectors)
{
    throw new UnsupportedRelationshipException();
}
```

Reply:
Worked like a charm! Thank you 😊